File: python/877/Stone_Game.py
class Solution(object):
    def stoneGame(self, piles):
        """
        :type piles: List[int]
        :rtype: bool
        """
        n = len(piles)
        dp = [[0 for _ in range(n)] for _ in range(n)]
        # dp[i][j] 表示[i,j]先拿者能赢的最大分数，dp[i][j]=max(piles[i]-dp[i+1][j], piles[j]-dp[i][j-1])
        # 先求子问题，l表示长度 1~n
        for i in range(n):
            dp[i][i] = piles[i]
        for l in range(2, n+1):
            i = 0
            j = i+l-1
            while j<n:
                dp[i][j] = max(piles[i]-dp[i+1][j], piles[j]-dp[i][j-1])
                i += 1
                j = i+l-1
        return dp[0][n-1] > 0

File: cpp/877/Stone_Game.cpp
class Solution {
public:
    bool stoneGame(vector<int>& piles) {
        vector<vector<int>> dp(piles.size(), vector<int>(piles.size(), 0));
        for(int i=0;i<piles.size();++i){
            dp[i][i] = piles[i];
        }
        for(int i=2;i<=piles.size();++i){
            for(int j=0;j+i-1<piles.size();++j){
                dp[j][j+i-1] = max(piles[j]-dp[j+1][j+i-1], piles[j+i-1]-dp[j][j+i-2]);
            }
        }
        return dp[0][piles.size()-1]>0;
    }
};